### PR TITLE
[Gateway] : One Gateway for all chains and deployments.

### LINKED ISSUE #211: Gateway for the multichain environment - task breakdown
The idea is to have the gateway look similar to: https://gnosis-safe.io/app/welcome
Where on the top right you can switch between different chain environments.
We would gradually provide inputs on what features are disabled/enabled on which chain.
For example:
Ecosystem subpage would have less items to show on Moonbeam/Fantom/Ethereum.

### LINKED ISSUE #422: [Gateway] UI UX Design to Multi chain
[Gateway] UI UX Design to Multi chain
Like gnosis safe.

### LINKED ISSUE #423: [Gateway] Prepare react component to down all supported chains update the application state.
[Gateway] Prepare react component to down all supported chains update the application state.

### LINKED ISSUE #424: [Gateway] Prepare the feature list to enable for each chain
[Gateway] Prepare the feature list to enable for each chain

### LINKED ISSUE #425: [Gateway] Extend network service to support multichain configuration base on user selection
[Gateway] Extend network service to support multichain configuration base on user selection

### LINKED ISSUE #459: [Gateway] : Features needs to be enable for each deployment on gateway
## Issue Type

```
[ ] bug report
[ x ] feature request
```

## Current Behavior
Currently we have all the features on gateway enable for the mainnet and rinkeby.

## Expected Behavior
Need help on what features needs to be enable for the each deployment like.

1. MoonBeam <> BobaBeam
2. MoonBase <> Bobabase
3. Avalanche <> Boba Avax
4. Fantom <> Boba Fantom

### PR DESCRIPTION
**Pull request by sk-enya**
:clipboard: Add associated issues, tickets, docs URL here.
closes: https://github.com/bobanetwork/boba/issues/459
closes: https://github.com/bobanetwork/boba/issues/425
closes: https://github.com/bobanetwork/boba/issues/424
closes: https://github.com/bobanetwork/boba/issues/423
closes: https://github.com/bobanetwork/boba/issues/422
closes: https://github.com/bobanetwork/boba/issues/211

## Overview

One gateway for all deployments or chains like on mainnet, testnet
**Ethereum, BNB, Fantom, Avax, Moonbeam, Moonbase.**

## Changes

Describe your changes and implementation choices. More details make PRs easier to review.

- Network Selector / Switcher to switch between the deployment or network from single gateway.
- Support added for networks [BNB, Etheruem, Fantom, Avax, BobaBeam].
- Add button `switch to selected network` once user change the network from dropdown above and reload the app with change.
- Initialise account on switching to network.
- Persisting the selected network in redux store so user won't loose previous selection on browser reload.
- Render the menu items based on selected network.
- BlockExplore URL on the footer should change w.r.to network and layer selection.
- Query param based network selection option `gateway.boba.network/bridge?chainId=XX` to select the network.
- Prepare separate network util along with with hook to read the query param data.
- Updated Sentry Packages to the latest.

## Testing
Describe how to test your new feature/bug fix and if possible, a step-by-step guide on how to demo this.

Refer screenshot visuals
Screenshot 2022-12-07 at 7 18 46 PM
Screenshot 2022-12-07 at 7 18 53 PM
Screenshot 2022-12-07 at 7 19 06 PM

### CODE REVIEW

File: packages/boba/gateway/src/components/mainMenu/layerSwitcher/LayerSwitcher.js [hunk 1]
```diff
@@ -247,10 +255,10 @@ function LayerSwitcher({
         aria-label="text alignment"
       >
         <ToggleButton sx={{p: "5px 10px", borderRadius: '12px 0 0 12px'}} value="L1" aria-label="L1">
-          <EthereumIcon selected={layer === 'L1'}/>
+          <L1Icon selected={layer === 'L1'}/>
         </ToggleButton>
         <ToggleButton sx={{p: "5px 10px", borderRadius: '0 12px 12px 0'}} value="L2" aria-label="L2">
-          <BobaIcon selected={layer === 'L2'} />
+          <L2Icon selected={layer === 'L2'} />
```
Comment (wsdt) [inline packages/boba/gateway/src/components/mainMenu/layerSwitcher/LayerSwitcher.js]: should we maybe create an enum or constant for these strings?

Comment (sk-enya) [reply packages/boba/gateway/src/components/mainMenu/layerSwitcher/LayerSwitcher.js]: you mean to say for  **layer** ?

File: packages/boba/gateway/src/containers/wallet/Wallet.js [hunk 2]
```diff
@@ -163,7 +163,7 @@ function Wallet() {
             }}
             variant="body2"
             component="span">
-            Ethereum Wallet
+            {networkName['l1'] || 'Etheruem'} Wallet
```
Comment (wsdt) [inline packages/boba/gateway/src/containers/wallet/Wallet.js]: *Ethereum
Maybe also convert the default values to a constant?

File: packages/boba/gateway/src/util/constant.js [hunk 3]
```diff
@@ -73,3 +72,13 @@ export const ROUTES_PATH = {
   DAO: '/DAO',
 }
 export const PER_PAGE = 8
+
+
+export const PAGES_BY_NETWORK = {
```
Comment (wsdt) [inline packages/boba/gateway/src/util/constant.js:74]: Have we actually ever thought about translating? 
Having a i18n json file giving it to some freelancer on Fiverr for a small fee and having a multi-lingual site etc. (only a small fraction of the world population speaks English)

File: packages/boba/gateway/src/util/network/config/moonbeam.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import { ETHERSCAN_API_KEY } from "util/constant";
+
+export const moonbeamConfig = {
```
Comment (wsdt) [inline packages/boba/gateway/src/util/network/config/moonbeam.js:1]: seeing these things makes me want to use TypeScript so much - not sure if we want to spend time on that, but it definitely would reduce the likelihood of bugs in future

**Review by wsdt — APPROVED** [review]
looks great, just left some comments!

File: packages/boba/gateway/src/components/mainMenu/layerSwitcher/LayerSwitcher.js (hunk 1 above)
Comment (wsdt) [reply packages/boba/gateway/src/components/mainMenu/layerSwitcher/LayerSwitcher.js]: Yes for `Layer.L1` or `Layer.L2`

File: packages/boba/gateway/src/util/constant.js (hunk 3 above)
Comment (sk-enya) [reply packages/boba/gateway/src/util/constant.js:74]: As of now, I don't know do we really it or not. but this is really awesome point good to have on gateway.. 
add issues for the same.
https://github.com/bobanetwork/boba/issues/490

File: packages/boba/gateway/src/util/network/config/moonbeam.js (hunk 4 above)
Comment (sk-enya) [reply packages/boba/gateway/src/util/network/config/moonbeam.js:1]: yeah, I am feeling the same as well.

Let me make note on issues and will take that up later. https://github.com/bobanetwork/boba/issues/491

**Comment by boyuan-chen** [comment]
Some problems:
1. The constant name of the default token should be fixed. For example, the default token name is `ETH` in `gateway/src/components/mainMenu/feeSwitcher/FeeSwitcher.js#L130`. In Alt L1s, the default token name is `BOBA`.
> We can store L1 token name, L1 token symbol in one js file and export them
2. The token addresses of L1 native token on L2 is not `0x4200000000000000000000000000000000000006`, this breaks the wallet page. The L2 address of L1 native token is `https://github.com/bobanetwork/boba/blob/844469e00d2618375128afa1789be6c464bea720/packages/boba/gateway/src/services/networkService.js#L122`.
> Maybe rewrite the logic of loading the default token?
3. The supported tokens on different networks are different. Maybe use this to fix it. `https://github.com/bobanetwork/boba/blob/844469e00d2618375128afa1789be6c464bea720/packages/boba/gateway/src/services/networkService.js#L197`
> Can we simplify this?
4. The function methods are different in `networkService.js`. `https://github.com/bobanetwork/boba/blob/alt-l1/packages/boba/gateway/src/services/networkService.js#L197`
> Maybe we import contract functions from another file, which automatically handles the function name and input? To avoid copying and pasting the `contract.json` file, you can use `new ethers.utils.Interface['XXX','XXX']` to load contracts.
```
new ethers.Contract(
  ADDRESS,
  new ethers.utils.Interface([
    'event TransactionEnqueued (address indexed _l1TxOrigin, address indexed _target, uint256 _gasLimit,bytes _data, uint256 indexed _queueIndex, uint256 _timestamp)'
  ]),
  wallet
)
```
I would suggest fixing these problems one by one.

**Comment by InoMurko** [comment]
is this up for review or are you missing things?

**Comment by sk-enya** [comment]
> is this up for review or are you missing things?

@InoMurko Yeah this is not yet ready I am trying to get the comments to be resolve which @Boyuan-Chen has added earlier above.

**Comment by InoMurko** [comment]
> > is this up for review or are you missing things?
> 
> 
> 
> @InoMurko Yeah this is not yet ready I am trying to get the comments to be resolve which @Boyuan-Chen has added earlier above.

Cool. Let us know!

**Comment by Boyuan-Chen** [comment]
BUG: 
1. wallet page doesn't work at all.
2. The frontend crashes.
image

**Comment by boyuan-chen** [comment]
The wallet page crashes. 
image

**Comment by boyuan-chen** [comment]
farm page crashes.
image

**Comment by boyuan-chen** [comment]
Other bugs: 
1. The `pool name` in the `earn` page should follow the chain name. Currently, they are named as `Ethereum Pools`.
2. `BOBA Chains` tab is broken.
3. The switch between `L2` and `L1` crashes the whole page.

**Comment by sk-enya** [comment]
> Other bugs:
> 
> 1. The `pool name` in the `earn` page should follow the chain name. Currently, they are named as `Ethereum Pools`.
> 2. `BOBA Chains` tab is broken.
> 3. The switch between `L2` and `L1` crashes the whole page.

@Boyuan-Chen Thank you for quick test and reply I fixed these issues and push change.

**Comment by sk-enya** [comment]
> farm page crashes. image

Please make sure we have to run the `yarn && yarn clean && yarn build` once you switch to this pr, as gateway dependencies get added correctly `redux-persist`

Anyways will try to reproduce and fix it.

**Review by InoMurko — APPROVED** [review]

**Review by wsdt — APPROVED** [review]

**Comment by boyuan-chen** [comment]
- The classic exit is broken on Goerli L2.
- Fee is not ETH on other L2s.
image

**Review by boyuan-chen — APPROVED** [review]

**Comment by souradeep-das** [comment]
Screenshot 2023-03-03 at 5 52 12 PM

came across this once - not sure if this is on my side
otherwise everything looks great to me! 🚀

**Review by souradeep-das — APPROVED** [review]

**Comment by alvaro-ricotta** [comment]
Hi guys, first comment on these changes, I was checking the repository and the new functionality looks amazing, I just have a few small suggestions/questions about these changes.

Shouldn't we incorporate Typescript in all our code whenever possible? Migrating the current code to Typescript shouldn't be a big effort, however the benefits of adding it are highly positive compared to the time it would take to implement it:

- Avoid data typing-related problems,
- Provide more clarity to development and be able to have complete tracking of the code we use at all times
- Adapt to industry standards.
- For integrations, although this may not be the case, when one develops in TS by default and imports libraries that do not have TS support, integration becomes a big problem as types must be manually set or ignored, which can generate major issues when connecting libraries together.

As I mentioned, these are just some improvements that I believe we should consider in the future. It is not required to migrate 100% to start, but it would be good to gradually implement it every time a developer modifies a product.

**Comment by sk-enya** [comment]
Hi @alvaro-ricotta,
thank you for sharing your thoughts
Yeah, Moving to typescript would be great improvement to the code base, as it help us with benefits which you mention above.

We already have issue for this which you can refer here and moving the comment to issue.
https://github.com/bobanetwork/boba/issues/491

**Pull request merged**